xds/internal/client/fakexds/fakexds.go

```diff
@@ -81,30 +78,35 @@ func StartClientAndServer(t *testing.T) (*Server, *grpc.ClientConn, func()) {
 	fs := &Server{
 		RequestChan:  make(chan *Request, defaultChannelBufferSize),
 		ResponseChan: make(chan *Response, defaultChannelBufferSize),
+		Address:      lis.Addr().String(),
 	}
 	adsgrpc.RegisterAggregatedDiscoveryServiceServer(server, fs)
 	go server.Serve(lis)
-	t.Logf("Starting fake xDS server at %v...", lis.Addr().String())
-	defer func() {
-		if err != nil {
-			server.Stop()
-			lis.Close()
-		}
-	}()
+	t.Logf("Starting fake xDS server at %v...", fs.Address)
+
+	return fs, func() {
+		lis.Close()
+		server.Stop()
+	}
+}
+
+// GetClientConn returns a grpc.ClientConn talking to the fake server. The
+// returned function should be invoked by the caller once the test is done.
+func (fs *Server) GetClientConn(t *testing.T) (*grpc.ClientConn, func()) {
+	t.Helper()
 
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 
-	var cc *grpc.ClientConn
-	cc, err = grpc.DialContext(ctx, lis.Addr().String(), grpc.WithInsecure(), grpc.WithBlock())
+	cc, err := grpc.DialContext(ctx, fs.Address, grpc.WithInsecure(), grpc.WithBlock())
 	if err != nil {
-		t.Fatalf("grpc.DialContext(%s) failed: %v", lis.Addr().String(), err)
+		t.Fatalf("grpc.DialContext(%s) failed: %v", fs.Address, err)
 	}
 	t.Log("Started xDS gRPC client...")
 
-	return fs, cc, func() {
-		server.Stop()
-		lis.Close()
+	return cc, func() {
+		cancel()
```

Comment:
Cancel is already called when this function returns.

Reply:
Done.